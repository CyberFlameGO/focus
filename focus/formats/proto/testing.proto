syntax = "proto3";

package testing;

message Foo {
	uint32 bar = 1;
}
