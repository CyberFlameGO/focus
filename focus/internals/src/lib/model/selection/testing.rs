use std::{collections::HashSet, path::Path, sync::Arc};

use focus_testing::{init_logging, ScratchGitRepo};

use anyhow::Result;
use focus_util::app::App;
use maplit::hashset;
use tempfile::TempDir;

use super::*;

struct Fixture {
    #[allow(dead_code)]
    dir: TempDir,
    underlying: ScratchGitRepo,
    app: Arc<App>,
}

impl Fixture {
    fn new() -> Result<Self> {
        let dir = TempDir::new()?;
        let path = dir.path().join("dense");
        let branch = String::from("main");
        let repo =
            ScratchGitRepo::new_copied_fixture(Path::new("bazel_java_example"), &path, &branch)?;
        let app = Arc::new(App::new(false)?);
        Ok(Self {
            dir,
            underlying: repo,
            app,
        })
    }

    fn repo(&self) -> Result<Repo> {
        Repo::open(self.underlying.path(), self.app.clone())
    }
}

/// Extract project names from the selection
fn project_names(selection: &Selection) -> HashSet<String> {
    selection
        .projects
        .iter()
        .map(|project| project.name.clone())
        .collect()
}

#[test]
fn repo_with_no_selections_returns_mandatory_projects_in_computed_selection() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let selections = repo.selections()?;
    let computed_selection = selections.computed_selection()?;
    assert_eq!(
        project_names(&computed_selection),
        hashset! {String::from("mandatory")}
    );
    assert!(computed_selection.targets.is_empty());

    Ok(())
}

#[test]
fn modifying_and_saving_the_selection() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let project_name = String::from("team_banzai/project_a");
    let target = Target::try_from("bazel://library_b/...")?;

    {
        let mut selections = repo.selections()?;
        selections.process(&vec![
            Operation {
                action: OperationAction::Add,
                underlying: Underlying::Project(project_name.clone()),
            },
            Operation {
                action: OperationAction::Add,
                underlying: Underlying::Target(target.clone()),
            },
        ])?;
        selections.save()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name.clone()}
        );
        assert_eq!(computed_selection.targets, hashset! {target.clone()});
    }

    {
        // Ensure that after loading from disk in a new instance, the selection is the same.
        let mut selections = repo.selections()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name.clone()}
        );
        assert_eq!(computed_selection.targets, hashset! {target.clone()});

        // Remove the target
        selections.process(&vec![Operation {
            action: OperationAction::Remove,
            underlying: Underlying::Target(target),
        }])?;
        selections.save()?;
        let computed_selection = selections.computed_selection()?;
        assert_eq!(
            project_names(&computed_selection),
            hashset! {String::from("mandatory"), project_name}
        );
        assert!(computed_selection.targets.is_empty());
    }

    Ok(())
}

#[test]
fn adding_an_unknown_project() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let nonexistent_project = Underlying::Project(String::from("blofeld/moonbase"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: nonexistent_project.clone(),
    }])?;
    assert_eq!(result.is_success(), false);
    assert_eq!(result.absent, hashset! {nonexistent_project});

    Ok(())
}

#[test]
fn mandatory_projects_cannot_be_selected() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let mandatory_project = Underlying::Project(String::from("mandatory"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: mandatory_project.clone(),
    }])?;
    assert_eq!(result.is_success(), false);
    assert_eq!(result.absent, hashset! {mandatory_project});

    Ok(())
}

#[test]
fn duplicate_projects_are_ignored() -> Result<()> {
    init_logging();

    let fixture = Fixture::new()?;
    let repo = fixture.repo()?;

    let project_b = Underlying::Project(String::from("team_zissou/project_b"));
    let mut selections = repo.selections()?;
    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: project_b.clone(),
    }])?;
    assert_eq!(result.is_success(), true);
    assert_eq!(result.added, hashset! {project_b.clone()});

    let result = selections.process(&vec![Operation {
        action: OperationAction::Add,
        underlying: project_b.clone(),
    }])?;
    assert_eq!(result.is_success(), true);
    assert_eq!(result.ignored, hashset! {project_b});

    Ok(())
}
