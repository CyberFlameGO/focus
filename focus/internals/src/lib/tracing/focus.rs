// module to collect tracing setup and config for the focus app itself

use std::io;

use anyhow::Result;
use tracing::dispatcher::DefaultGuard;
use tracing_appender::non_blocking::WorkerGuard;
use tracing_error::ErrorLayer;
use tracing_subscriber::prelude::*;
use tracing_subscriber::{self, util::SubscriberInitExt, EnvFilter};

#[derive(Debug)]
pub enum GuardWrapper {
    WorkerGuard(WorkerGuard),
    DefaultGuard(DefaultGuard),
}

#[derive(Debug, Default)]
/// opaque struct for returning tracing WorkerGuard instances to main
pub struct Guard {
    _inner: Vec<GuardWrapper>,
}

impl From<WorkerGuard> for Guard {
    fn from(wg: WorkerGuard) -> Self {
        Guard {
            _inner: vec![GuardWrapper::WorkerGuard(wg)],
        }
    }
}

#[derive(Debug, Default)]
pub struct TracingOpts {
    pub is_tty: bool,
    pub nocolor_requested: bool,
}

pub fn init_tracing(opts: TracingOpts) -> Result<Guard> {
    let TracingOpts {
        is_tty,
        nocolor_requested,
    } = opts;
    let use_color = is_tty && !nocolor_requested;
    let (non_blocking, guard) = tracing_appender::non_blocking(io::stderr());

    tracing_subscriber::registry()
        .with(ErrorLayer::default())
        .with(EnvFilter::new(
            std::env::var("RUST_LOG").unwrap_or_else(|_| "info".to_string()),
        ))
        .with(
            tracing_subscriber::fmt::layer()
                .with_span_events(tracing_subscriber::fmt::format::FmtSpan::CLOSE)
                .with_target(false)
                .with_writer(non_blocking)
                .with_ansi(use_color),
        )
        .try_init()?;

    Ok(Guard::from(guard))
}
