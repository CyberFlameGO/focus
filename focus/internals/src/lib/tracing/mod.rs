pub mod chrome;
pub mod git_trace2;

pub use chrome::Trace;

#[cfg(test)]
pub(in crate::tracing) mod testing {
    use std::path::{Path, PathBuf};

    use anyhow::{anyhow, Result};

    pub const FIXTURE_RELPATH: &str = "src/lib/tracing/fixtures";

    pub fn manifest_dir() -> Result<PathBuf> {
        Ok(PathBuf::from(
            std::env::var_os("CARGO_MANIFEST_DIR").ok_or(anyhow!("CARGO_MANIFEST_DIR not set"))?,
        ))
    }

    pub fn manifest_relative_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        Ok(manifest_dir()?.join(s.as_ref()))
    }

    pub fn fixture_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        manifest_relative_path(FIXTURE_RELPATH).map(|p| p.join(s.as_ref()))
    }
}
