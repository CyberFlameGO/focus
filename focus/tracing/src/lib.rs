// Copyright 2022 Twitter, Inc.
// SPDX-License-Identifier: Apache-2.0

pub mod chrome;
pub mod focus;
pub mod git_trace2;

use std::path::{Path, PathBuf};

pub use crate::focus::{init_tracing, Guard, TracingOpts};
pub use chrome::Trace;

use anyhow::Result;

#[cfg(test)]
pub(crate) mod testing {
    use std::path::{Path, PathBuf};

    use anyhow::{anyhow, Result};

    pub const FIXTURE_RELPATH: &str = "src/fixtures";

    pub fn manifest_dir() -> Result<PathBuf> {
        Ok(PathBuf::from(
            std::env::var_os("CARGO_MANIFEST_DIR")
                .ok_or_else(|| anyhow!("CARGO_MANIFEST_DIR not set"))?,
        ))
    }

    pub fn manifest_relative_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        Ok(manifest_dir()?.join(s.as_ref()))
    }

    pub fn fixture_path<S: AsRef<Path>>(s: S) -> Result<PathBuf> {
        manifest_relative_path(FIXTURE_RELPATH).map(|p| p.join(s.as_ref()))
    }
}

fn home_relative_path<P: AsRef<Path>>(p: P) -> Result<PathBuf> {
    match dirs::home_dir().map(|pb| pb.join(p.as_ref())) {
        Some(path) => Ok(path),
        None => Err(anyhow::anyhow!("HOME not defined")),
    }
}

#[cfg(target_os = "macos")]
const DEFAULT_LOG_DIR: &str = "Library/Logs/focus";

#[cfg(not(any(
    target_os = "windows",
    target_os = "macos",
    target_os = "ios",
    target_arch = "wasm32"
)))]
const DEFAULT_LOG_DIR: &str = ".local/focus/log";

/// returns the default system specific log location
pub fn log_dir() -> Result<PathBuf> {
    home_relative_path(DEFAULT_LOG_DIR)
}
