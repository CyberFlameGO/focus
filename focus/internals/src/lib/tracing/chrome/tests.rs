#![cfg(test)]

use super::*;
use crate::tracing::testing::*;
use anyhow::Result;

fn assert_trace_snapshot<S: AsRef<str>>(input: S) -> Result<()> {
    let trace = Trace::git_trace_from_file(&fixture_path(input.as_ref())?)?;
    insta::assert_yaml_snapshot!(input.as_ref(), trace);
    Ok(())
}

#[test]
fn test_clone_focus_trace() -> Result<()> {
    assert_trace_snapshot("clone-perf.json")
}

#[test]
fn test_status_trace() -> Result<()> {
    assert_trace_snapshot("status.json")
}

#[test]
fn test_push_trace() -> Result<()> {
    assert_trace_snapshot("push.json")
}

#[test]
fn test_reset_trace() -> Result<()> {
    assert_trace_snapshot("reset.json")
}

#[test]
fn test_reflog_trace() -> Result<()> {
    assert_trace_snapshot("reflog.json")
}

#[test]
fn test_rebase_trace() -> Result<()> {
    assert_trace_snapshot("rebase.json")
}

#[test]
fn test_branch_update_trace() -> Result<()> {
    assert_trace_snapshot("branch-update.json")
}
