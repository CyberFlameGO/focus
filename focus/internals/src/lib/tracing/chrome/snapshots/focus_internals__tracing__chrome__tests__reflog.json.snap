---
source: focus/internals/src/lib/tracing/chrome/tests.rs
assertion_line: 12
expression: trace

---
traceEvents:
  - ts: 0
    pid: 1
    tid: 0
    name: main
    cat: ""
    ph: X
    dur: 21072650
    args:
      begin:
        file_line_thread: "common-main.c:49:main"
        git_event_name: version
        git_version: 2.35.1
        stats_version: "3"
      end:
        code: 0
        file_line_thread: "trace2/tr2_tgt_event.c:201:main"
        git_event_name: atexit
        t_abs: 21.073792
  - ts: 56
    pid: 1
    tid: 0
    name: git reflog
    cat: ""
    ph: X
    dur: 31273
    args:
      begin:
        argv:
          - git
          - reflog
        file_line_thread: "common-main.c:50:main"
        git_event_name: start
        t_abs: 0.001196
      end:
        code: 0
        file_line_thread: "git.c:716:main"
        git_event_name: exit
        t_abs: 0.03247
  - ts: 330
    pid: 1
    tid: 0
    name: "repository.c:136:main"
    cat: ""
    ph: i
    args:
      file_line_thread: "repository.c:136:main"
      git_event_name: def_repo
      worktree: /Users/jsimms/workspace/focus
  - ts: 501
    pid: 1
    tid: 0
    name: "git.c:459:main"
    cat: ""
    ph: i
    args:
      file_line_thread: "git.c:459:main"
      git_event_name: cmd_name
      hierarchy: reflog
      name: reflog
  - ts: 5106
    pid: 1
    tid: 0
    name: less
    cat: ""
    ph: X
    dur: 21067480
    args:
      begin:
        argv:
          - less
        cd: ~
        child_class: pager
        child_id: 0
        file_line_thread: "run-command.c:721:main"
        git_event_name: child_start
        hook_name: ~
        use_shell: true
      end:
        code: 0
        file_line_thread: "run-command.c:977:main"
        git_event_name: child_exit
        pid: 31194
        t_rel: 21.067475
displayTimeUnit: ~
systemTraceEvents: ~
otherData: ~

